### DISCUSSION TITLE
How to connect line to non-rectangle/non-circle shape (each shape lives inside a square in debug mode)?

### DISCUSSION THREAD

**Discussion by Andrew15-5** [cetz-package/cetz#156] [Q&A] [answered]
I have a manual operation block in a flowchart. I need to connect lines from the sides to `op.left` and `op.right`, but since the block is inside a rectangle in debug mode, lines don't connect to the center of the left and right edges. Is there any way to solve this? Or is it a limitation/bug?

![image](https://github.com/johannes-wolf/typst-canvas/assets/37143421/99c0a250-bc23-4327-bca3-fe41f1c780b6) ![image](https://github.com/johannes-wolf/typst-canvas/assets/37143421/ff4bcaac-22b2-401c-9e34-e31399948ba5)

```typ
#let manual-operation(name, start: none, direction: "down", text) = {
  import cetz.draw: *
  validate-start-coordinate(start)
  validate-direction-vector(direction, valid-directions: ("down", "right"))
  if direction == "down" {
    start = (rel: (0, -height / 2), to: start)
  } else if direction == "right" {
    start = (rel: (width / 2, 0), to: start)
  }
  line(
    (rel: (-width / 2, height / 2), to: start),
    (rel: (width, 0)),
    (rel: (-width / 4, -height)),
    (rel: (-width / 4 * 2, 0)),
    close: true,
    name: name,
  )
  content(name + ".center", text)
}
```

P.S. The right line should connect to the right edge of the block, but the left line isn't really connecting to anything anyway (this is kinda a bug in my implementation for now).

---

**Update**:

I was able to fix it by manually calculating perimeter and all edges, then using `place-anchors()` to override `.left` and `.right` with precise ratios/offsets that represent center of left/right edges:

```typ
  ...
  let top = width
  let bottom = width / 2
  let side = calc.sqrt(calc.pow((top - bottom) / 2, 2) + calc.pow(height, 2))
  let perimeter = top + bottom + side * 2
  let right-offset = (top + side / 2) / perimeter
  let left-offset = (top + side + bottom + side / 2) / perimeter
  place-anchors(name: name, line(
    (rel: (-width / 2, height / 2), to: start),
    (rel: (width, 0)),
    (rel: (-width / 4, -height)),
    (rel: (-width / 4 * 2, 0)),
    close: true,
    name: name,
    // ), (name: "left", pos: .88), (name: "right", pos: .465))
  ), (name: "left", pos: left-offset), (name: "right", pos: right-offset))
  if direction == "right" { line(start-old, name + ".left") }
  ...
```

![image](https://github.com/johannes-wolf/typst-canvas/assets/37143421/5007af2f-f85f-4a70-83e7-185a6b250694) ![image](https://github.com/johannes-wolf/typst-canvas/assets/37143421/5ea25c2b-1dc7-4e00-95f4-59898105e0cc)

This technically answers the question, but this is way more code/calculations that in my mind should be done. The anchor positions should be correct automatically or by providing a flag such as `compute-anchor-positions` if there is a potential performance hit (be lazy by default).

**Comment by johannes-wolf** [comment, marked as answer]
I would do it by placing the path in a named group and overwriting the left/right anchor:
```
#let manual-operation(name, start: none, direction: "down", text) = {
  import cetz.draw: *
  validate-start-coordinate(start)
  validate-direction-vector(direction, valid-directions: ("down", "right"))
  if direction == "down" {
    start = (rel: (0, -height / 2), to: start)
  } else if direction == "right" {
    start = (rel: (width / 2, 0), to: start)
  }
  group(name: name, {
    let (tl, tr, bl, br) = ((rel: (-width / 2, +height / 2), to: start),
                            (rel: (+width / 2, +height / 2), to: start),
                            (rel: (-width / 4, -height / 2), to: start),
                            (rel: (+width / 4, -height / 2), to: start))
    line(tl, tr, br, bl, close: true)
    anchor("left", (tl, .5, bl))
    anchor("right", (tr, .5, br))
  })
  content(name + ".center", text)
}
```

If you calculate all 4 coordinates, you can use linear interpolation `(a, f, b)` between `a` and `b` to get the center of an edge. The `left`, `right`, ... anchors default to the (transformed) bounding box of an element.

**Reply by Andrew15-5** [reply]
Yes! This is what I wanted to do, but somehow I didn't figure out a way to use/apply interpolation coordinates. This is a very clean solution. Thanks!
